providers/base/tests/test_wwan_tests.py

```diff
@@ -109,3 +111,98 @@ def test_invoked_with_mmdbus(self, mock_mmdbus):
             self.assertTrue(mmdbus_instance.get_model_name.called)
             self.assertTrue(mmdbus_instance.get_firmware_revision.called)
             self.assertTrue(mmdbus_instance.get_hardware_revision.called)
+
+
+class TestThreeGppScanTest(unittest.TestCase):
+
+    def test_register_argument(self):
+
+        sys.argv = ["wwan_tests.py", "3gpp-scan", "2", "--timeout", "600"]
+        obj_3gppscan = wwan_tests.ThreeGppScanTest()
+        ret_args = obj_3gppscan.register_argument()
+        self.assertEqual(ret_args.hw_id, "2")
+        self.assertEqual(ret_args.timeout, 600)
+
+    @patch("subprocess.check_call")
+    @patch("wwan_tests._wwan_radio_off")
+    @patch("wwan_tests._wwan_radio_on")
+    @patch("wwan_tests.MMCLI")
+    @patch("wwan_tests.ThreeGppScanTest.register_argument")
+    def test_invoked_successfully(
+        self, mock_arg, mock_mmcli, mock_r_on, mock_r_off, mock_subprocess
+    ):
+        mock_arg.return_value = argparse.Namespace(hw_id="2", timeout=200)
+
+        mmcli_instance = Mock()
+        mmcli_instance.equipment_id_to_mm_id.return_value = "0"
+        mock_mmcli.return_value = mmcli_instance
+        mock_subprocess.return_value = 0
+
+        with self.assertRaises(SystemExit) as context:
+            obj_3gppscan = wwan_tests.ThreeGppScanTest()
+            obj_3gppscan.invoked()
+
+        self.assertEqual(context.exception.code, 0)
+        self.assertTrue(mock_mmcli.called)
+        self.assertTrue(mmcli_instance.equipment_id_to_mm_id.called)
+        self.assertTrue(mock_r_on.called)
+        self.assertTrue(mock_r_off.called)
+        mock_subprocess.assert_called_with(
+            ["mmcli", "-m", "0", "--3gpp-scan", "--timeout", "200"]
+        )
+
+    @patch("subprocess.check_call")
+    @patch("wwan_tests._wwan_radio_off")
+    @patch("wwan_tests._wwan_radio_on")
+    @patch("wwan_tests.MMCLI")
+    @patch("wwan_tests.ThreeGppScanTest.register_argument")
+    def test_invoked_failed(
+        self, mock_arg, mock_mmcli, mock_r_on, mock_r_off, mock_subprocess
+    ):
+        mock_arg.return_value = argparse.Namespace(hw_id="2", timeout=200)
+
+        mmcli_instance = Mock()
+        mmcli_instance.equipment_id_to_mm_id.return_value = "0"
+        mock_mmcli.return_value = mmcli_instance
+        mock_subprocess.return_value = 1
+
+        with self.assertRaises(SystemExit) as context:
+            obj_3gppscan = wwan_tests.ThreeGppScanTest()
+            obj_3gppscan.invoked()
+
+        self.assertEqual(context.exception.code, 1)
+        self.assertTrue(mock_mmcli.called)
+        self.assertTrue(mmcli_instance.equipment_id_to_mm_id.called)
+        self.assertTrue(mock_r_on.called)
+        self.assertTrue(mock_r_off.called)
+        mock_subprocess.assert_called_with(
+            ["mmcli", "-m", "0", "--3gpp-scan", "--timeout", "200"]
+        )
```

Comment:
Given that you also test when a `CalledProcessError` is raised, I don't think this test is required, as it covers the same parts of your code.

(even if you remove it, the coverage for `wwan_tests.py` will still be 100%)

Reply:
no, the coverage will down to 89%. So that's why I added this unit test.

Reply:
here are some error symptom while scanning 3GPP network. So we could simply check if the `error` message exists or not.
```
# the modem is not enabled
admin@C0310A2043-00843:~$ sudo mmcli -m 1 --3gpp-scan --timeout 200; echo $?
error: modem not enabled yet
1

# Failed to scan available network within 100s
admin@C0310A2043-00843:~$ sudo mmcli -m 0 --3gpp-scan --timeout 100; echo $?
error: couldn't scan networks in the modem: 'Timeout was reached'
1
```